### DISCUSSION TITLE
Authentication Recommendation

### DISCUSSION THREAD

**Discussion by SueHeir** [nandorojo/solito#94] [Q&A]
Hello people!

So I'm new to React, NextJS, React Native, and now Solito.  Last month I learned how to make a website in NextJS with authentication, profiles, posts, upvotes, etc.  I am now trying to learn app development and came across this project but there isn't much in terms of tutorials or guides. 

I am wanting to know what is a recommended way to do Authentication with Solito that would work well with both React Native and NextJS.  I used Cookies with my first website project, and found this might be an option, [react-native-cookies](https://github.com/react-native-cookies/cookies), but am looking for opinions and suggestions on what would work, and how to get it to work. (i.e do they need to be done separately or can they be done together within solito)

**Comment by nandorojo** [comment]
hey there!

so…cookie based auth probably won’t work here, since cookies are a web-only concept. at least, that’s my understanding.

there is an open discussion about auth at #39

i use firebase auth on both web and native and it works great. it’s JWT-based. i want to add a solito example, but it’s a good amount of work out of the many requests I get, so I haven’t had time to quite yet. 

the easiest thing may be would be to use firebase JS. depends what auth providers you’d want I think.

i use firebase JS on web and react-native-firebase on iOS/android for optimal native performance. without a good starter template, this might be a bit more setup for a beginner. 

clearly auth is a big need for an example / documentation, and hopefully when i free up some time from work i can make an example with what we do.

**Comment by SueHeir** [comment]
So I am going to give firebase a try! I have been following [rnfirebase.io](https://rnfirebase.io/) (toward the end of getting started where they talk about expo) and I ran into an error 

```
⚠️  Something went wrong running `pod install` in the `ios` directory.
Command `pod install` failed.
└─ Cause: The following Swift pods cannot yet be integrated as static libraries:

The Swift pod `FirebaseCoreInternal` depends upon `GoogleUtilities`, which does not define modules. To opt into those targets generating module maps (which is necessary to import them from Swift when building as static libraries), you may set `use_modular_headers!` globally in your Podfile, or specify `:modular_headers => true` for particular dependencies.
```

And fixed it with adding use_modular_headers! to the pod file

Expo now runs but when I run the app I get the following errors

```
Invariant Violation: Native module cannot be null.
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/react-native/Libraries/Core/ExceptionsManager.js:104:6 in reportException
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/react-native/Libraries/Core/ExceptionsManager.js:172:19 in handleException
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/react-native/Libraries/Core/setUpErrorHandling.js:24:6 in handleError
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/@react-native/polyfills/error-guard.js:49:36 in ErrorUtils.reportFatalError
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/metro-runtime/src/polyfills/require.js:204:6 in guardedLoadModule
at http://192.168.1.226:19000/index.bundle?platform=ios&dev=true&hot=false&strict=false&minify=false:188795:3 in global code

Invariant Violation: "main" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called.
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/react-native/Libraries/Core/ExceptionsManager.js:104:6 in reportException
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/react-native/Libraries/Core/ExceptionsManager.js:172:19 in handleException
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/react-native/Libraries/Core/setUpErrorHandling.js:24:6 in handleError
at /Users/elizabethsuehr/Documents/GitHub/debaccle-app/node_modules/@react-native/polyfills/error-guard.js:49:36 in ErrorUtils.reportFatalError
```


I understand if you want to wait until you release an example first before helping people with this, but I can't seem to find anything online to help with this error. Any pointers on how you did it would be appreciated!  Im assuming you added  "@react-native-firebase/app": "^15.1.1" to the package.json in the expo folder and then ran expo prebuild --clean. I'll be willing to write up a complete guide if I ever get it working lol :)

**Reply by nandorojo** [reply]
since it uses custom native code, you need to use a custom Dev Client instead of expo go (`expo run:ios`). once you have that, you can use `expo start —dev-client`. i recommend looking at expo’s docs about custom dev clients

**Reply by SueHeir** [reply]
Sweet I got it working at least on expo run:ios, thank you!

**Reply by SueHeir** [reply]
One problem though it that npx create-solito-app@latest my-solito-app used react-native-reanimated  version~2.3.1 , and to get the ios pods to compile i needed to use version 2.9.1. Maybe this can be updated?

**Reply by nandorojo** [reply]
yeah looks like the solito starter needs to `expo upgrade` to SDK 45. can be kinda tricky in a monorepo, but another thing on my list…

**Comment by SueHeir** [comment]
### Setting Up Firebase and expo-auth-session with Solito

(Please comment on problems you find or any suggestions!)

First, install Solito with  
```
npx create-solito-app@latest my-solito-app && cd my-solito-app/apps/expo
```
Now install FireBase and expo-auth-session. (I believe expo-auth-session can be used for google logins with firebase but I haven't confirmed this, I will do soon)

```
expo install @react-native-firebase/app @react-native-firebase/perf @react-native-firebase/crashlytics @react-native-firebase/auth expo-auth-session expo-random 
```
Also, you will need to update react-native-reanimated to version 2.8.0.  then run
```
expo install
```
Note: upgrading to 2.9.0 breaks nextjs

Edit your app.json file to include the following, Also now would be a good time to set up your Firebase Project and add an ios and android app to your Firebase project to get the following files: GoogleService-Info.plist, google-services.json. These files need to be added to the root directory of /expo before you continue. Note "com.my-solito-app" need to match what you entered on the Firebase website.  
```json
 "ios": {
      "bundleIdentifier": "com.my-solito-app",
      "googleServicesFile": "./GoogleService-Info.plist"
    },
    "android": {
      "package": "com.my-solito-app",
      "googleServicesFile": "./google-services.json"
    },
    "plugins": [
      "@react-native-firebase/app",
      "@react-native-firebase/perf",
      "@react-native-firebase/crashlytics"
    ]
```

Now let us run the following command, this will fail because ios pods aren't set up correctly, but it will do a lot of the work for us in terms of setting up firebase in ios and android
```
expo prebuild --clean
```



Now to set up the podfile correctly with the following lines
```
  use_modular_headers!
  use_frameworks! :linkage => :static

```


After that, you can clean and install pods
```
cd ios
rm -rf ~/Library/Developer/Xcode/DerivedData
rm -rf Pod 
pod deintegrate && pod cache clean --all 
pod install --repo-update 
```


Now we can build the ios or android
```
expo run:ios
```


It should be installed now, to test this add the following into a firebase.tsx file, this will be accompanied by a firebase.web.tsx file later.
```js
import auth from '@react-native-firebase/auth'

export function createUser() {
  auth()
    .createUserWithEmailAndPassword(
      'jane.doe@example.com',
      'SuperSecretPassword!'
    )
    .then(() => {
      console.log('User account created & signed in!')
    })
    .catch((error) => {
      if (error.code === 'auth/email-already-in-use') {
        console.log('That email address is already in use!')
      }

      if (error.code === 'auth/invalid-email') {
        console.log('That email address is invalid!')
      }

      console.error(error)
    })
}
```
And add the following to any screen to test!
```js
import { createUser } from 'app/provider/FirebaseAuth/firebase'

//later in file
<MotiPressable onPress={createUser}>
      <Text> Click to create Account</Text>
</MotiPressable>
```


You should see a console.log of 'User account created & signed in!' if it's your first time. **Note** you need to setup username-and-password signup on the firebase website under authentication -> Sign-in Methods

### Setup Nextjs 

Go into /apps/next and run 
```
yarn add firebase
```
to test if this is working add the following into _app.tsx. The firebaseConfig info comes from the firebase website
```js
import { initializeApp } from 'firebase/app'

const firebaseConfig = {
  apiKey: '',
  authDomain: '',
  projectId: '',
  storageBucket: '',
  messagingSenderId: '',
  appId: '',
  measurementId: '',
}

// Initialize Firebase
const app = initializeApp(firebaseConfig)
```
And then add the following into the file firebase.web.tsx
```js
import { getAuth, createUserWithEmailAndPassword } from 'firebase/auth'

export function createUser() {
  const auth = getAuth()
  createUserWithEmailAndPassword(
    auth,
    'jane.doe@example.com',
    'SuperSecretPassword!'
  )
    .then(() => {
      console.log('User account created & signed in!')
    })
    .catch((error) => {
      if (error.code === 'auth/email-already-in-use') {
        console.log('That email address is already in use!')
      }

      if (error.code === 'auth/invalid-email') {
        console.log('That email address is invalid!')
      }

      console.error(error)
    })
}
```
When calling this function you should get an error saying this email is already in use!

I believe this covers everything that you would need to use Firebase Auth with Solito!

**Reply by cloudorbush** [reply]
thank you for sharing :)

**Reply by juanchaur** [reply]
@SueHeir thanks for sharing.
I'm following your steps but I'm getting this error:

```bash
❌  (/app/apps/expo/ios/Pods/FirebaseCoreInternal/FirebaseCore/Internal/Sources/HeartbeatLogging/RingBuffer.swift:19:35)

  17 | /// Error types for `RingBuffer` operations.
  18 | enum RingBufferError: LocalizedError {
> 19 |   case outOfBoundsPush(pushIndex: Array.Index, endIndex: Array.Index)
     |                                   ^ reference to generic type 'Array' requires arguments in <...>
  20 | 
  21 |   var errorDescription: String {
  22 |     switch self {

› 5 error(s), and 0 warning(s)

CommandError: Failed to build iOS project. "xcodebuild" exited with error code 65.
error Command failed with exit code 1.

```
Did it happen to you too?
